### ISSUE TITLE
Qualifiers in intersection types

### ISSUE THREAD

**Issue by lnsun** [typetools/checker-framework#3349]
The qualifiers in intersecton types is confusing me by allowing two exclusive qualifiers in one type. For example:

```
import org.checkerframework.checker.nullness.qual.*;
import java.io.Serializable;

class MyClass<T extends @NonNull Object & @Nullable Serializable> {
    void foo(T p1){
        @NonNull Serializable s = p1;  // OK, but checker issues warning
        @NonNull Object o = p1;
    }
}
```
The first assignment operation of above code yields this error:
```
Error: [assignment.type.incompatible] incompatible types in assignment.
   found   : T extends @Initialized @NonNull Object & @Initialized @Nullable Serializable
   required: @UnknownInitialization @NonNull Serializable
```
That indicates `T` have both qualifiers, but `T` is only one type and cannot be instantiated with `@Nullable` arguments. So, it is quite confusing to have the error.

**Comment by lnsun** [comment]
As @wmdietl point out, maybe the greatest lower bound of all qualifiers should be used as the only qualifier of the intersection type.

**Comment by mernst** [comment]
Here are two action items:
 * For this issue, ensure that type-checking uses the GLB of all the type qualifiers.  The false positive is because type-checking uses just the qualifier written on one of the conjuncts.
 * As a separate issue (#3362), warn when the behavior may be surprising.
